### DISCUSSION TITLE
Invalid hook calls for SB v7 + webpack 5 + react 18.2.0 + pnpm

### DISCUSSION THREAD

**Discussion by ZhengRui** [storybookjs/storybook#22650] [Help]
I am writing a react lib with "react" and "react-dom" as peer dependency, then I have a seperate storybook project demoing the usage of the lib. It used to work well with `SB v6.5 + react 17/18 + webpack5 + yarn`, recently i upgraded to `SB v7 + react 18 + webpack5 + pnpm`, and i am facing the issue of not being able to call any react hooks in the lib. 

The original [comment](https://github.com/storybookjs/storybook/issues/11904#issuecomment-1555449716) i make on this [issue](https://github.com/storybookjs/storybook/issues/11904), and here is [a minimum reproduction](https://github.com/ZhengRui/sbv7-issue-repro) of my situation. @shilman

**Comment by ZhengRui** [comment]
able to find the fix: remove `react` and `react-dom` from lib, only keep `@types/react` and `pnpm link` react to the storybook project's react.

**Reply by ZhengRui** [reply]
seems it still conflicts with zustand. whenever I call "useStore" in the lib, storybook throws error "Cannot read properties of null (reading 'useRef')"

**Reply by ZhengRui** [reply]
seems i have to treat zustand just like react, install zustand on the storybook project and then link it in the lib. This brings a big issue: I have to do this for all the dependencies that uses react hooks. Not sure if this is an issue of storybook or an issue of pnpm, but it's super annoying.

**Reply by ZhengRui** [reply]
btw, reversly linking react/react-dom/zustand from lib to storybook project doesn't work, it also has "invalid hook call" error.

**Reply by ZhengRui** [reply]
give up the idea of making the lib and sb as two separate projects, just make them a combo single project to save all the hassles

**Reply by ZhengRui** [reply]
now comes another issue, when putting libA and storybook in one project libA's components works fine, but if libA's component imports a component from libB, then storybook will throw "invalid hook call" for libB. this is so frustrating!

**Reply by bigbunty** [reply]
Does this work now? Iam also facing similar issue with SB7 +react18.2.0+webpack5
<img width="1173" alt="Screenshot 2023-07-03 at 4 13 31 PM" src="https://github.com/storybookjs/storybook/assets/108940398/dfdbb4d2-6c63-45ea-9e18-96f6ff1f0309">

**Reply by liaoyinglong** [reply]
After adding the following code to `.storybook/main.js`, it was fixed.

```js
webpackFinal(config){
 config.resolve.alias = {
      ...config.resolve.alias,
      react: path.resolve(__dirname, '../node_modules/react'),
      'react-dom': path.resolve(__dirname, '../node_modules/react-dom'),
    };
return config
}

```

**Reply by felipeegasneris** [reply]
+1 @liaoyinglong solution

**Reply by AndrewSepic** [reply]
+1 @liaoyinglong   🙏

**Comment by sky247-app** [comment]
Sorry to hear that you're facing issues with calling React hooks in your React library after upgrading to Storybook v7, React 18, webpack 5, and pnpm. Here is a suggested comment for the issue:

I've been working on a React library that has "react" and "react-dom" as peer dependencies. Previously, everything was functioning well with Storybook v6.5, React 17/18, webpack 5, and yarn. However, I recently upgraded to Storybook v7, React 18, webpack 5, and pnpm, and now I'm encountering an issue where I'm unable to call any React hooks within the library.

I've prepared a minimal reproduction of the situation to help investigate the problem. It seems like there may be a compatibility issue update on [Crecimiento organico](https://crecimientoorganico.com.mx) between the upgraded dependencies and the hooks within the library.

Has anyone else experienced a similar issue or found a workaround? Any insights or suggestions would be greatly appreciated. Thank you!
